test/Microsoft.AspNetCore.OData.Tests/Query/Expressions/QueryBinderTests.cs

```diff
@@ -100,6 +106,214 @@ public void BindSingleResourceFunctionCallNode_ThrowsNotSupported_ForNotAcceptNo
+
+        // Create an instance of QueryBinderContext using the real model and settings
+        Type clrType = model.GetClrType(employeeType);
+        var context = new QueryBinderContext(model, new ODataQuerySettings(), clrType);
+
+        // Act
+        Expression expression = binder.BindSingleResourceFunctionCallNode(node, context);
+
+        // Assert
+        Assert.NotNull(expression);
+        // cast($it, 'Microsoft.AspNetCore.OData.Tests.Models.Manager') where $it is the Employee entity
+        Assert.Equal("($it As Manager)", expression.ToString());
+        Assert.Equal("Microsoft.AspNetCore.OData.Tests.Models.Manager", expression.Type.ToString());
+        Assert.Equal(typeof(Manager), expression.Type);
+        Assert.Equal(ExpressionType.TypeAs, expression.NodeType);
+        Assert.Equal(employeeType.FullName(), (expression as UnaryExpression).Operand.Type.FullName);
+        Assert.Equal(managerType.FullName(), (expression as UnaryExpression).Type.FullName);
+    }
+
+    public static TheoryDataSet<List<QueryNode>> BindSingleResourceFunctionCallNodeForPropertyCasting_Data
+    {
+        get
+        {
+            var addressType = HardCodedTestModel.GetEdmComplexTypeFor("Microsoft.AspNetCore.OData.Tests.Models.Address");
+            var workAddressType = HardCodedTestModel.GetEdmComplexTypeFor("Microsoft.AspNetCore.OData.Tests.Models.WorkAddress");
+
+            // Create a ResourceRangeVariableReferenceNode for the Employee entity
+            var employeeNode = CreateEmployeeRangeVariableReferenceNode();
+
+            // Create a SingleComplexNode for the Location property of the Person entity
+            var locationProperty = HardCodedTestModel.GetEdmPropertyFor("Microsoft.AspNetCore.OData.Tests.Models.Employee", "Location");
+            var locationNode = new SingleComplexNode(employeeNode as SingleResourceNode, locationProperty);
+
+            int capacity = 2;
+            // Create a ConstantNode to cast Location to NS.WorkAddress
+            var parametersWithConstantNode = new List<QueryNode>(capacity)
+            {
+                // First parameter is the Location property
+                locationNode, 
+                // Second parameter is the ConstantNode
+                new ConstantNode("Microsoft.AspNetCore.OData.Tests.Models.WorkAddress") 
+            };
+
+            // Create a SingleResourceCastNode to cast Location to Microsoft.AspNetCore.OData.Tests.Models.WorkAddress
+            var parametersWithSingleResourceCastNode = new List<QueryNode>(capacity)
+            {
+                locationNode,
+                new SingleResourceCastNode(locationNode, workAddressType) 
+            };
+
+            var data = new TheoryDataSet<List<QueryNode>>()
+            {
+                { parametersWithConstantNode },
+                { parametersWithSingleResourceCastNode }
+            };
+
+            return data;
```

Comment:
In my opinion, it would be easier to read and understand this data set if the data set were defined in a single pass, so you don't have to hold too much context in your head, e.g.:

```suggestion
        return new TheoryDataSet<List<QueryNode>>()
        {
            {
                locationNode,
                new ConstantNode("Microsoft.AspNetCore.OData.Tests.Models.WorkAddress")
            },
            {
                locationNode,
                new SingleResourceCastNode(locationNode, workAddressType)
            }
        }
```